src/contract_wrappers/contract_wrapper.ts

```diff
@@ -1,13 +1,46 @@
 import * as _ from 'lodash';
 import * as Web3 from 'web3';
 import {Web3Wrapper} from '../web3_wrapper';
-import {ZeroExError, Artifact} from '../types';
+import {AbiDecoder} from '../utils/abi_decoder';
+import {
+    ZeroExError,
+    Artifact,
+    LogWithDecodedArgs,
+    RawLog,
+    ContractEvents,
+    SubscriptionOpts,
+    IndexedFilterValues,
+} from '../types';
 import {utils} from '../utils/utils';
 
 export class ContractWrapper {
     protected _web3Wrapper: Web3Wrapper;
-    constructor(web3Wrapper: Web3Wrapper) {
+    private _abiDecoder?: AbiDecoder;
+    constructor(web3Wrapper: Web3Wrapper, abiDecoder?: AbiDecoder) {
         this._web3Wrapper = web3Wrapper;
+        this._abiDecoder = abiDecoder;
+    }
+    protected async _getLogsAsync(address: string, eventName: ContractEvents, subscriptionOpts: SubscriptionOpts,
+                                  indexFilterValues: IndexedFilterValues,
+                                  abi: Web3.ContractAbi): Promise<LogWithDecodedArgs[]> {
+        // TODO include indexFilterValues in topics
+        const eventSignature = this._getEventSignatureFromAbiByName(abi, eventName);
+        const filter = {
+            fromBlock: subscriptionOpts.fromBlock,
+            toBlock: subscriptionOpts.toBlock,
+            address,
+            topics: [this._web3Wrapper.keccak256(eventSignature)],
+        };
+        const logs = await this._web3Wrapper.getLogsAsync(filter);
+        const logsWithDecodedArguments = _.map(logs, this._tryToDecodeLogOrNoOp.bind(this));
+        return logsWithDecodedArguments;
+    }
+    protected _tryToDecodeLogOrNoOp(log: Web3.LogEntry): LogWithDecodedArgs|RawLog {
+        if (_.isUndefined(this._abiDecoder)) {
+            throw new Error(ZeroExError.NoAbiDecoder);
```

Comment:
Hm... I think we want to reserve the `ZeroExError` type of errors that the user might encounter. Since this error will only ever be seen by devs, let's start a new error type called `InternalZeroExError`.